File: ai/registry.ts
/**
 * Agent registry - configurations for different agent types.
 *
 * Defines agent configs for different use cases:
 * - build: Primary agent with full tool access
 * - general: Subagent for parallel task execution
 * - explore: Read-only for codebase search
 * - plan: Analysis and planning without file writes
 */

import type { AgentToolName } from './tools';

export type AgentMode = 'primary' | 'subagent';

export interface AgentConfig {
  name: string;
  description: string;
  mode: AgentMode;
  systemPrompt: string;
  temperature: number;
  topP: number;
  /** Tools enabled for this agent (true = enabled, false = disabled) */
  toolsEnabled: Partial<Record<AgentToolName | string, boolean>>;
  /** Shell command patterns allowed (glob patterns) */
  allowedShellPatterns?: string[];
}

// Default system prompt prefix
const SYSTEM_PROMPT_PREFIX = `You are an AI assistant helping with software development tasks.
You have access to tools for reading, writing, and searching files.

Key principles:
- Always read files before modifying them
- Make minimal, targeted changes
- Explain your reasoning before making changes
- Ask for clarification when requirements are unclear
`;

// Agent configurations
const buildAgent: AgentConfig = {
  name: 'build',
  description: 'Primary agent with full tool access for development tasks',
  mode: 'primary',
  systemPrompt: `${SYSTEM_PROMPT_PREFIX}
You are the primary development agent. You have full access to all tools
and can read, write, and modify files as needed to complete tasks.

You can interact with GitHub using the github tool for:
- Creating and viewing pull requests
- Cloning and syncing repositories
- Creating and viewing issues
- Checking CI/CD status
`,
  temperature: 0.7,
  topP: 0.95,
  toolsEnabled: {
    grep: true,
    readFile: true,
    writeFile: true,
    multiedit: true,
    webFetch: true,
    github: true,
  },
  allowedShellPatterns: ['*'], // All commands allowed
};

const generalAgent: AgentConfig = {
  name: 'general',
  description: 'Subagent for parallel task execution',
  mode: 'subagent',
  systemPrompt: `${SYSTEM_PROMPT_PREFIX}
You are a general-purpose subagent executing a specific task.
Focus on completing your assigned objective efficiently.
`,
  temperature: 0.7,
  topP: 0.95,
  toolsEnabled: {
    grep: true,
    readFile: true,
    writeFile: true,
    multiedit: true,
    webFetch: true,
    github: true,
  },
  allowedShellPatterns: ['*'],
};

const exploreAgent: AgentConfig = {
  name: 'explore',
  description: 'Read-only agent for fast codebase exploration',
  mode: 'subagent',
  systemPrompt: `${SYSTEM_PROMPT_PREFIX}
You are an exploration agent focused on searching and understanding code.
You can only read files and search - you cannot modify anything.
Be thorough but efficient in your search.
`,
  temperature: 0.5,
  topP: 0.9,
  toolsEnabled: {
    grep: true,
    readFile: true,
    writeFile: false,
    multiedit: false,
    webFetch: false,
  },
  allowedShellPatterns: [
    'ls *',
    'find *',
    'tree *',
    'git log *',
    'git show *',
    'git diff *',
    'git status',
  ],
};

const planAgent: AgentConfig = {
  name: 'plan',
  description: 'Analysis and planning agent (read-only)',
  mode: 'subagent',
  systemPrompt: `${SYSTEM_PROMPT_PREFIX}
You are a planning agent. Analyze the codebase and create implementation plans.
You can read files and search but cannot modify anything.
Focus on understanding architecture and proposing clear, actionable plans.
`,
  temperature: 0.6,
  topP: 0.9,
  toolsEnabled: {
    grep: true,
    readFile: true,
    writeFile: false,
    multiedit: false,
    webFetch: true, // Can fetch documentation
  },
  allowedShellPatterns: [
    'ls *',
    'find *',
    'tree *',
    'git *',
    'cat *',
  ],
};

// Registry map
const agentRegistry = new Map<string, AgentConfig>([
  ['build', buildAgent],
  ['general', generalAgent],
  ['explore', exploreAgent],
  ['plan', planAgent],
]);

/**
 * Get agent configuration by name.
 */
export function getAgentConfig(name: string): AgentConfig {
  const config = agentRegistry.get(name);
  if (!config) {
    // Fall back to build agent
    return buildAgent;
  }
  return config;
}

/**
 * Check if a tool is enabled for an agent.
 */
export function isToolEnabled(agentName: string, toolName: string): boolean {
  const config = getAgentConfig(agentName);
  const enabled = config.toolsEnabled[toolName as AgentToolName];
  // Default to true if not specified
  return enabled !== false;
}

/**
 * Check if a shell command is allowed for an agent.
 */
export function isShellCommandAllowed(agentName: string, command: string): boolean {
  const config = getAgentConfig(agentName);
  const patterns = config.allowedShellPatterns;

  if (!patterns || patterns.length === 0) {
    return false;
  }

  // Check for wildcard pattern
  if (patterns.includes('*')) {
    return true;
  }

  // Simple glob matching
  for (const pattern of patterns) {
    if (matchGlob(pattern, command)) {
      return true;
    }
  }

  return false;
}

/**
 * Simple glob pattern matching.
 */
function matchGlob(pattern: string, text: string): boolean {
  // Convert glob to regex
  const regexPattern = pattern
    .replace(/[.+^${}()|[\]\\]/g, '\\$&') // Escape special chars
    .replace(/\*/g, '.*') // * -> .*
    .replace(/\?/g, '.'); // ? -> .

  const regex = new RegExp(`^${regexPattern}$`, 'i');
  return regex.test(text);
}

/**
 * List all registered agent names.
 */
export function listAgentNames(): string[] {
  return Array.from(agentRegistry.keys());
}

/**
 * Register a custom agent configuration.
 */
export function registerAgent(config: AgentConfig): void {
  agentRegistry.set(config.name, config);
}

export { buildAgent, generalAgent, exploreAgent, planAgent };

File: server-zig/src/ai/registry.zig
const std = @import("std");
const types = @import("types.zig");

const AgentConfig = types.AgentConfig;
const AgentMode = types.AgentMode;

/// System prompt prefix used by all agents
const SYSTEM_PROMPT_PREFIX =
    \\You are an AI assistant helping with software development tasks.
    \\You have access to tools for reading, writing, and searching files.
    \\
    \\Key principles:
    \\- Always read files before modifying them
    \\- Make minimal, targeted changes
    \\- Explain your reasoning before making changes
    \\- Ask for clarification when requirements are unclear
;

/// Build agent - primary agent with full tool access
pub const build_agent = AgentConfig{
    .name = "build",
    .description = "Primary agent with full tool access for development tasks",
    .mode = .primary,
    .system_prompt = SYSTEM_PROMPT_PREFIX ++
        \\
        \\You are the primary development agent with full access to:
        \\- File operations (read, write, edit)
        \\- Code search (grep)
        \\- Command execution (PTY)
        \\- Web access
        \\- GitHub CLI
        \\
        \\Use these tools carefully and always verify your changes.
    ,
    .temperature = @as(f32, 0.7),
    .top_p = @as(f32, 0.95),
    .tools_enabled = .{},
    .allowed_shell_patterns = &.{"*"},
};

/// Explore agent - read-only for fast codebase exploration
pub const explore_agent = AgentConfig{
    .name = "explore",
    .description = "Read-only agent for fast codebase exploration",
    .mode = .subagent,
    .system_prompt = SYSTEM_PROMPT_PREFIX ++
        \\
        \\You are an exploration agent with read-only access.
        \\Your job is to quickly search and analyze code without making changes.
        \\
        \\Focus on:
        \\- Finding relevant files and code patterns
        \\- Understanding code structure
        \\- Summarizing what you find
        \\
        \\You cannot write or modify files.
    ,
    .temperature = @as(f32, 0.5),
    .top_p = @as(f32, 0.9),
    .tools_enabled = .{
        .grep = true,
        .read_file = true,
        .write_file = false,
        .multiedit = false,
        .web_fetch = false,
        .github = true,
        .unified_exec = false,
        .write_stdin = false,
        .close_pty_session = false,
        .list_pty_sessions = false,
    },
    .allowed_shell_patterns = &.{ "ls *", "find *", "tree *", "git log *", "git show *", "git diff *", "git status" },
};

/// Plan agent - analysis and planning (read-only)
pub const plan_agent = AgentConfig{
    .name = "plan",
    .description = "Analysis and planning agent (read-only)",
    .mode = .subagent,
    .system_prompt = SYSTEM_PROMPT_PREFIX ++
        \\
        \\You are a planning agent focused on analysis and design.
        \\
        \\Your job is to:
        \\- Analyze requirements
        \\- Understand existing code structure
        \\- Design implementation approaches
        \\- Create detailed plans
        \\
        \\You can read files and fetch web content but cannot modify code.
    ,
    .temperature = @as(f32, 0.6),
    .top_p = @as(f32, 0.9),
    .tools_enabled = .{
        .grep = true,
        .read_file = true,
        .write_file = false,
        .multiedit = false,
        .web_fetch = true,
        .github = true,
        .unified_exec = false,
        .write_stdin = false,
        .close_pty_session = false,
        .list_pty_sessions = false,
    },
    .allowed_shell_patterns = &.{ "ls *", "find *", "tree *", "git *", "cat *" },
};

/// General agent - full access subagent
pub const general_agent = AgentConfig{
    .name = "general",
    .description = "General-purpose subagent with full tool access",
    .mode = .subagent,
    .system_prompt = SYSTEM_PROMPT_PREFIX ++
        \\
        \\You are a general-purpose agent that can be spawned for parallel tasks.
        \\You have full access to all tools.
        \\
        \\Complete your assigned task efficiently and report results clearly.
    ,
    .temperature = @as(f32, 0.7),
    .top_p = @as(f32, 0.95),
    .tools_enabled = .{},
    .allowed_shell_patterns = &.{"*"},
};

/// Get agent configuration by name
pub fn getAgentConfig(name: []const u8) AgentConfig {
    if (std.mem.eql(u8, name, "build")) return build_agent;
    if (std.mem.eql(u8, name, "explore")) return explore_agent;
    if (std.mem.eql(u8, name, "plan")) return plan_agent;
    if (std.mem.eql(u8, name, "general")) return general_agent;
    return build_agent; // Default fallback
}

/// Check if a tool is enabled for an agent
pub fn isToolEnabled(agent_name: []const u8, tool_name: []const u8) bool {
    const config = getAgentConfig(agent_name);

    if (std.mem.eql(u8, tool_name, "grep")) return config.tools_enabled.grep;
    if (std.mem.eql(u8, tool_name, "readFile")) return config.tools_enabled.read_file;
    if (std.mem.eql(u8, tool_name, "writeFile")) return config.tools_enabled.write_file;
    if (std.mem.eql(u8, tool_name, "multiedit")) return config.tools_enabled.multiedit;
    if (std.mem.eql(u8, tool_name, "webFetch")) return config.tools_enabled.web_fetch;
    if (std.mem.eql(u8, tool_name, "github")) return config.tools_enabled.github;
    if (std.mem.eql(u8, tool_name, "unifiedExec")) return config.tools_enabled.unified_exec;
    if (std.mem.eql(u8, tool_name, "writeStdin")) return config.tools_enabled.write_stdin;
    if (std.mem.eql(u8, tool_name, "closePtySession")) return config.tools_enabled.close_pty_session;
    if (std.mem.eql(u8, tool_name, "listPtySessions")) return config.tools_enabled.list_pty_sessions;

    return true; // Default to enabled
}

/// List all available agent names
pub const AGENT_NAMES: []const []const u8 = &.{
    "build",
    "explore",
    "plan",
    "general",
};

test "getAgentConfig returns correct config" {
    const build = getAgentConfig("build");
    try std.testing.expectEqualStrings("build", build.name);
    try std.testing.expectEqual(AgentMode.primary, build.mode);

    const explore = getAgentConfig("explore");
    try std.testing.expectEqualStrings("explore", explore.name);
    try std.testing.expectEqual(AgentMode.subagent, explore.mode);
}

test "isToolEnabled respects agent config" {
    // Build agent should have all tools enabled
    try std.testing.expect(isToolEnabled("build", "writeFile"));
    try std.testing.expect(isToolEnabled("build", "grep"));

    // Explore agent should not have write tools
    try std.testing.expect(!isToolEnabled("explore", "writeFile"));
    try std.testing.expect(isToolEnabled("explore", "grep"));
}
